app/MMB/src/list/FlightList.js

```diff
@@ -0,0 +1,274 @@
+        |},
+      |},
+    |},
+  |},
+|};
+
+type MulticityType = {|
+  +status: ?string,
+  +databaseId: ?number,
+  +end: ?{|
+    +airport: ?{|
+      +city: ?{|
+        +name: ?string,
+      |},
+    |},
+  |},
+  +start: ?{|
+    +time: ?any,
+    +airport: ?{|
+      +city: ?{|
+        +name: ?string,
+      |},
+    |},
+  |},
+|};
+
+type FlightDetails = {|
+  departureDate: ?Date,
+  departureCity: ?string,
+  status: ?string,
+  bookingId: ?number,
+  finalDestination: ?string,
+  passengerCount: ?number,
+  imageUrl: ?string,
+  type: ?string,
+|};
+
+type Props = {|
+  data: FlightListType,
+|};
+
+export class FlightList extends React.Component<Props> {
+  getFlightDetails = (): FlightDetails[] => {
+    const flights = idx(this.props.data, _ => _.allBookings.edges) || [];
+    return flights.reduce((acc, current) => {
+      const type = idx(current, _ => _.node.type) || '';
+      const baseDetails = {
+        type,
+        imageUrl: idx(current, _ => _.node.destinationImageUrl),
+        passengerCount: idx(current, _ => _.node.passengerCount),
+      };
+      switch (type) {
+        case 'ONE_WAY':
+          return [
+            ...acc,
+            {
+              ...baseDetails,
+              ...this.oneWayToFlightDetail(idx(current, _ => _.node.oneWay)),
+            },
+          ];
+        case 'RETURN':
+          return [
+            ...acc,
+            {
+              ...baseDetails,
+              ...this.returnToFlightDetail(idx(current, _ => _.node.return)),
+            },
+          ];
+        case 'MULTICITY':
+          return [
+            ...acc,
+            {
+              ...baseDetails,
+              ...this.multicityToFlightDetail(
+                idx(current, _ => _.node.multicity),
+              ),
+            },
+          ];
+        default:
+          return acc;
+      }
+    }, []);
+  };
+
+  oneWayToFlightDetail = (flight: ?OneWayType) => {
+    return {
+      departureDate: idx(flight, _ => _.trip.departure.time),
+      departureCity: idx(flight, _ => _.trip.departure.airport.city.name),
+      status: idx(flight, _ => _.status),
+      bookingId: idx(flight, _ => _.databaseId),
+      finalDestination: idx(flight, _ => _.trip.arrival.airport.city.name),
+    };
+  };
+
+  returnToFlightDetail = (flight: ?ReturnType) => ({
+    departureDate: idx(flight, _ => _.outbound.departure.time),
+    departureCity: idx(flight, _ => _.outbound.departure.airport.city.name),
+    status: idx(flight, _ => _.status),
+    bookingId: idx(flight, _ => _.databaseId),
+    finalDestination: idx(flight, _ => _.outbound.arrival.airport.city.name),
+  });
+
+  multicityToFlightDetail = (flight: ?MulticityType) => ({
+    departureDate: idx(flight, _ => _.start.time),
+    departureCity: idx(flight, _ => _.start.airport.city.name),
+    status: idx(flight, _ => _.status),
+    bookingId: idx(flight, _ => _.databaseId),
+    finalDestination: idx(flight, _ => _.end.airport.city.name),
+  });
```

Comment:
Is it possible to split it into 3 components? It would be great even from the Relay fragment perspective (this is actually how we designed the graph). I would use this component only for the decision which one to render. Makes sense?

Reply:
Yes, and it looks a lot nicer 🙂